File: src/vm/mod/iodev/test/slow_ints.cpp
#include "slow_ints.hpp"

namespace kcpu::iodev {

slow_ints::slow_ints(pic_in_interface &pic) : single_port_io_device(PORT_BASE), pic(pic) {
}

std::pair<regval_t, halfcycle_count_t> slow_ints::read() {
    throw vm_error("unimplemented");
}

halfcycle_count_t slow_ints::write(regval_t val) {
    if(!(val & MASK_NMI_FLAG)) {
        count[0] = (val & ~MASK_NMI_FLAG) + 1;
    }

    if(val & MASK_NMI_FLAG) {
        count[1] = (val & ~MASK_NMI_FLAG) + 1;
    }

    return 0;
}

void slow_ints::single_count_process_halfcycle(int count_num, int int_num) {
    if(!count[count_num]) {
        return;
    }

    count[count_num]--;

    if(!count[count_num]) {
        pic.assert(int_num);
    }
}

void slow_ints::process_halfcycle(bool offclock) {
    single_count_process_halfcycle(0, INT_NUM);
    single_count_process_halfcycle(1, NMI_NUM);
}

}

File: crate/kcpu/src/vm/io/dev/test/slow_ints.rs
use super::super::pic::Pic;
use crate::spec::types::hw::*;
use crate::vm::interface;
use crate::vm::io::types::*;

const PORT_BASE: Word = 0xD1;

pub struct SlowInts {
    pic: Handle<Pic>,

    count: [Word; 2],
}

impl SlowInts {
    const MASK_NMI_FLAG: Word = 0x8000;
    const NMI_NUM: u8 = 0;
    const INT_NUM: u8 = 3;

    pub fn new(pic: Handle<Pic>) -> Self {
        SlowInts { pic, count: [0, 0] }
    }

    fn process_halfcycle_register(&mut self, num: usize, int_num: u8) {
        if self.count[num] == 0 {
            return;
        }

        self.count[num] -= 1;

        if self.count[num] == 0 {
            interface::Pic::assert(&self.pic, interface::PicIrq::Num(int_num));
        }
    }
}

impl SinglePortDevice for SlowInts {
    fn reserved_port(&self) -> Word {
        PORT_BASE
    }

    fn write(&mut self, val: Word) -> HalfcycleCount {
        if val & SlowInts::MASK_NMI_FLAG == 0 {
            self.count[0] = (val & !SlowInts::MASK_NMI_FLAG) + 1;
        }

        if val & SlowInts::MASK_NMI_FLAG != 0 {
            self.count[1] = (val & !SlowInts::MASK_NMI_FLAG) + 1;
        }

        0
    }

    fn read(&mut self) -> (HalfcycleCount, Word) {
        panic!("read from SlowInts register");
    }

    fn process_halfcycle(&mut self, _: ClockedSignals) {
        self.process_halfcycle_register(0, SlowInts::NMI_NUM);
        self.process_halfcycle_register(1, SlowInts::INT_NUM);
    }
}
